src/LinearSolverTypes_Multigrid.f90

```diff
@@ -0,0 +1,973 @@
+      PROCEDURE,PASS :: init => init_LinearSolverType_Multigrid
+      !> @copybrief LinearSolverType_Multigrid::preAllocPETScInterpMat_LinearSolverType_Multigrid
+      !> @copydetails LinearSolverType_Multigrid::preAllocPETScInterpMat_LinearSolverType_Multigrid
+      PROCEDURE,PASS :: preAllocPETScInterpMat => &
+                          preAllocPETScInterpMat_LinearSolverType_Multigrid
+      !> @copybrief LinearSolverType_Multigrid::setupPETScMG_LinearSolverType_Multigrid
+      !> @copydetails LinearSolverType_Multigrid::setupPETScMG_LinearSolverType_Multigrid
+      PROCEDURE,PASS :: setupPETScMG => setupPETScMG_LinearSolverType_Multigrid
+      !> @copybrief LinearSolverType_Multigrid::fillInterpMats_LinearSolverType_Multigrid
+      !> @copydetails LinearSolverType_Multigrid::fillInterpMats_LinearSolverType_Multigrid
+      PROCEDURE,PASS :: fillInterpMats => fillInterpMats_LinearSolverType_Multigrid
+      !> @copybrief  LinearSolverType_Multigrid::solve_LinearSolverType_Multigrid
+      !> @copydetails LinearSolverType_Multigrid::solve_LinearSolverType_Multigrid
+      PROCEDURE,PASS :: solve => solve_LinearSolverType_Multigrid
+      !> @copybrief  LinearSolverType_Multigrid::clear_LinearSolverType_Multigrid
+      !> @copydetails LinearSolverType_Multigrid::clear_LinearSolverType_Multigrid
+      PROCEDURE,PASS :: clear => clear_LinearSolverType_Multigrid
+      !> @copybrief  LinearSolverType_Multigrid::setSmoother_LinearSolverType_Multigrid
+      !> @copydetails LinearSolverType_Multigrid::setSmoother_LinearSolverType_Multigrid
+      PROCEDURE,PASS :: setSmoother => setSmoother_LinearSolverType_Multigrid
+      !> @copybrief  LinearSolverType_Multigrid::updatedA_LinearSolverType_Multigrid
+      !> @copydetails LinearSolverType_Multigrid::updatedA_LinearSolverType_Multigrid
+      PROCEDURE,PASS :: updatedA => updatedA_LinearSolverType_Multigrid
+      !> @copybrief  LinearSolverType_Multigrid::setInterp_LinearSolverType_Multigrid
+      !> @copydetails LinearSolverType_Multigrid::setInterp_LinearSolverType_Multigrid
+      PROCEDURE,PASS :: setInterp => setInterp_LinearSolverType_Multigrid
+  ENDTYPE LinearSolverType_Multigrid
+
+  !> Logical flag to check whether the required and optional parameter lists
+  !> have been created yet for the Linear Solver Type.
+  LOGICAL(SBK),SAVE :: LinearSolverType_Paramsflag=.FALSE.
+
+  !> Name of module
+  CHARACTER(LEN=*),PARAMETER :: modName='LINEARSOLVERTYPES_MULTIGRID'
+!
+!===============================================================================
+  CONTAINS
+!
+!-------------------------------------------------------------------------------
+!> @brief Initializes the multigrid Linear Solver Type with a parameter list
+!>
+!> @param solver The linear solver to act on
+!> @param Params the parameter list
+!> @param A The A operator in Ax=b
+!>
+    SUBROUTINE init_LinearSolverType_Multigrid(solver,Params,A)
```

Comment:
Seems like some duplicate code here compared with init_LinearSolverType_Base.  Might be worth exploring moving the similar code into a true initBase method.  Probably doesn't have to be done right now though.  I'm not sure of the extent of the duplication.

Reply:
There is quite a bit of duplicate code, but I couldn't think of a better way to implement it in the standard LS types without putting in a bunch of long IF-blocks.  The downside is that if someone wants to change something common to both inits, they would have to do it twice.  I probably won't have time to implement it in a more optimal manner before I leave, but, if someone thinks of a good way to refactor this, that would be good!